generator/camera_info_collector.cpp

```diff
@@ -0,0 +1,343 @@
+    m2::RectD const rect =
+      MercatorBounds::RectByCenterXYAndSizeInMeters(m_data.m_center, kSearchCameraRadiusMeters);
+    auto bestSegment = polyline.UpdateProjection(rect);
+    double tmpCoef = 0;
+
+    if (bestSegment.IsValid())
+    {
+      static double constexpr kEps = 1e-6;
+      auto const & p1 = polyline.GetPolyline().GetPoint(bestSegment.m_ind);
+      auto const & p2 = polyline.GetPolyline().GetPoint(bestSegment.m_ind + 1);
+
+      if (AlmostEqualAbs(p1, p2, kEps))
+        return;
+
+      m2::ParametrizedSegment<m2::PointD> st(p1, p2);
+      auto const cameraProjOnSegment = st.ClosestPointTo(m_data.m_center);
+      curMinDist = MercatorBounds::DistanceOnEarth(cameraProjOnSegment, m_data.m_center);
+
+      tmpCoef = MercatorBounds::DistanceOnEarth(p1, cameraProjOnSegment) /
+                MercatorBounds::DistanceOnEarth(p1, p2);
+    }
+
+    if (curMinDist < minDist)
+    {
+      minDist = curMinDist;
+      bestFeatureId = ft.GetID().m_index;
+      segmentIdOfBestFeatureId = static_cast<uint32_t>(bestSegment.m_ind);
+      coef = tmpCoef;
+    }
+  };
+
+  dataSource.ForEachInRect(
+    updClosestFeatureCallback,
+    MercatorBounds::RectByCenterXYAndSizeInMeters(m_data.m_center, kSearchCameraRadiusMeters),
+    scales::GetUpperScale());
+
+  if (minDist != kMaxDistFromCameraToClosestSegmentMeters)
+    m_data.m_ways.emplace_back(bestFeatureId, segmentIdOfBestFeatureId, coef);
+}
+
+boost::optional<std::pair<double, uint32_t>> CamerasInfoCollector::Camera::FindMyself(
+    uint32_t wayFeatureId, FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId) const
+{
+  double coef = 0.0;
+  bool isRoad = true;
+  uint32_t result = 0;
+
+  auto const readFeature = [&](FeatureType & ft) {
+    bool found = false;
+    isRoad = routing::IsRoad(feature::TypesHolder(ft));
+    if (!isRoad)
+      return;
+
+    auto const forEachPoint = [&result, &found, this](m2::PointD const & pt) {
+      if (found)
+        return;
+
+      if (AlmostEqualAbs(m_data.m_center, pt, kCoordEqualityEps))
+        found = true;
+      else
+        ++result;
+    };
+
+    ft.ForEachPoint(forEachPoint, scales::GetUpperScale());
+
+    CHECK(found, ("Cannot find camera point at the feature:", wayFeatureId,
+                  "; camera center:", MercatorBounds::ToLatLon(m_data.m_center)));
+
+    // If point with number - N, is end of feature, we cannot say: segmentId = N,
+    // because number of segments is N - 1, so we say segmentId = N - 1, and coef = 1
+    // which means, that camera placed at the end of (N - 1)'th segment.
+    if (result == ft.GetPointsCount() - 1)
+    {
+      CHECK_NOT_EQUAL(result, 0, ("Feature consists of one point!"));
+      --result;
+      coef = 1.0;
+    }
+  };
+
+  FeatureID featureID(mwmId, wayFeatureId);
+  dataSource.ReadFeature(readFeature, featureID);
+
+  if (isRoad)
+    return boost::optional<std::pair<double, uint32_t>>({coef, result});
```

Comment:
Здесь получается, что coef 0 или 1. Что делать с камерами, которые находятся на сегментах, но не совпадают с концами? Судя по документации к методу, этот случай тоже должен обрабатываться.

Reply:
Нет, в доке сказано про камеры, центр которых располагается на концах сегментов:
```cpp
// Otherwise returns id of segment from feature with id - |wayId|, which starts at camera's
// center and coefficient - where it placed at the segment.
```

Reply:
Для камер, которые не совпадает с концами, есть другой метод:
```
FindClosestSegmentWithGeometryIndex
```